src/node_buffer.cc

```diff
@@ -752,12 +753,23 @@ uint32_t FastByteLengthUtf8(Local<Value> receiver,
   if (source.length > 128) {
     return simdutf::utf8_length_from_latin1(source.data, source.length);
   }
+
   uint32_t length = source.length;
   uint32_t result = length;
-  const uint8_t* data = reinterpret_cast<const uint8_t*>(source.data);
-  for (uint32_t i = 0; i < length; ++i) {
+  const auto data = reinterpret_cast<const uint8_t*>(source.data);
+
+  uint32_t i = 0;
+  const auto length8 = length & ~0x7;
+  while (i < length8) {
+    result += std::popcount(*reinterpret_cast<const uint64_t*>(data + i) & 0x8080808080808080);
```

Comment:
You can count on this function to be correct, I think.

But it might be technically undefined behaviour because data + i might not be aligned. (This does not actually matter on most systems.)

Performance-wise the function `std::popcount` won't necessarily be compiled to a population count instruction... on x64 because it is often not assumed, and on ARM because there is no standard (non-SIMD) popcount. So you will be likely getting some autovectorization going in some cases, and other cases, you will get a function call (possibly with runtime dispatching).


(My comments are *not* requests for change.)

Reply:
> But it might be technically undefined behaviour because data + i might not be aligned. (This does not actually matter on most systems.)

Fixed.

> Performance-wise the function std::popcount won't necessarily be compiled to a population count instruction... on x64 because it is often not assumed, and on ARM because there is no standard (non-SIMD) popcount. So you will be likely getting some autovectorization going in some cases, and other cases, you will get a function call (possibly with runtime dispatching).

Some compilers generate bit fiddling code ([divide an conqour popcount](https://github.com/hcs0/Hackers-Delight/blob/master/popArrayHS.c.txt) that is still faster than a loop.